### ISSUE TITLE
Question: Batched Events or Go Channels or Both

### ISSUE THREAD

**Issue by skyzyx** [newrelic/newrelic-client-go#460]
Looking at https://pkg.go.dev/github.com/newrelic/newrelic-client-go@v0.40.0/pkg/events?tab=doc#pkg-overview and I don't see any examples or guidance for when it makes sense to do this vs something else. I don't have any streams available to me (on input), so I'm fetching a bunch of data from one service, and pushing it into New Relic as custom events.

1. Trying to be as fast/efficient as possible, I'm already using Go Channels and I'm looking at the [`Events.BatchMode`](https://pkg.go.dev/github.com/newrelic/newrelic-client-go@v0.40.0/pkg/events?tab=doc#Events.BatchMode) method and wondering if there is an example of this usage.

2. Also, does it make more sense to use Go Channels to create several Batches to run in parallel, or is it better to just push everything into a single batch?

3. Lastly, does this work like a _pool_, where I can keep throwing more data into an existing batch, and the SDK will just keep pushing the data as fast as it can? Or do I need to batch it up, stop, send the batch and wait for a response, then re-open a new batch to accept more data?

**Comment by skyzyx** [comment]
Digging through the source and found this: https://github.com/newrelic/newrelic-client-go/blob/master/pkg/events/example_batch_test.go

**Comment by zlesnr** [comment]
Briefly looking at the code, I think there is a high water mark sort of notion, that when the number of events in the cache is reached, then they are all posted together.

https://github.com/newrelic/newrelic-client-go/blob/master/pkg/events/events_batch.go#L162

Since each batch sender has its own water mark, to spin up multiple of them would mean that you have the potential of more data sitting in the cache that has not been sent.  Given that, I think that using a single batch sender would make the most sense.  The flush in the test you pointed at is used when shutting down to make sure the batch gets sent.  So call that in the event that you need to shutdown your process to ensure all events get sent.

**Comment by skyzyx** [comment]
Appreciated. Can any notes on this topic and some example code be added to the GoDoc? Should I create a separate ticket for that?

**Comment by zlesnr** [comment]
Sure thing.  A ticket for this in particular would be good, and maybe the handful of things you would have expected to see in the doc.  I can read the code over and put up a PR for it.  Too, if you run into any issues let us know, or if you have code examples to throw our way, feel free to do so.
